probe-rs/src/session.rs

```diff
@@ -43,7 +43,7 @@ pub struct Session {
 
 #[derive(Debug)]
 enum ArchitectureInterface {
-    Arm(Box<dyn ArmProbeInterface>),
+    Arm(Box<dyn ArmProbeInterface + Send + Sync + 'static>),
```

Comment:
Why is this change necessary?

Reply:
That fors not matter. Every dupertype will only ever be Send + Sync if all itd dubtypes are too :)
So if this here is not it will propagate upwards until Session :)

Reply:
But why does `Session` have to be Send + Sync? Does `cargo-embed` need this?

Reply:
Ah, Yep :D I mean, to actually use it concurrently with the RTT runner and the GDB runner, one has to run on a separate thread :) Atm they sleep 10ms after they are done and thus give the other time :) Ofc, this can be done by multiplexing work on both with manual calls but then the API surface gets way more difficult :)

Reply:
I agree that Send is necessary for that, but I think `Sync` is not necessary, if we usually wrap the whole `Session` in a Mutext.

I would also prefer if the `Send` requirement would be placed directly on the `ArmProbeInterface`, since it doesn't really make sense to have one which is not `Send`. And then we don't have to specify it everywhere manually :wink:

Reply:
Yeah, `Sync` is not really necessary I believe. But I think it's rare to make a structure yourself, that is `Send` but not `Sync` :)

I can move it to the `ArmProbeInterface`, sounds sensible :)

Reply:
It's probably more idiomatic to require both `Send` and `Sync`, that's true. If possible, I would also like it if there was some kind of test in probe-rs itself which fails when `Session` is not Send + Sync, otherwise we only notice when cargo-embed (or something else) fails.

Reply:
> If possible, I would also like it if there was some kind of test in probe-rs itself which fails when `Session` is not Send + Sync, otherwise we only notice when cargo-embed (or something else) fails.

I agree this would be great! maybe something like:

```
let session = Session::auto_attach(())?;
std::thread::spawn(move || session.core(0););
```

As an example in the docs is enough already to ensure this :)

Reply:
I like the new test for `Send + Sync`! 👍 

I think you forgot to move the `Send + Sync` bound to the `ArmProbeInterface` itself, however.